relay-server/src/services/buffer/mod.rs

```diff
@@ -81,6 +84,44 @@ impl FromMessage<Self> for EnvelopeBuffer {
     }
 }
 
+/// Abstraction that wraps a list of [`ObservableEnvelopeBuffer`]s to which [`Envelope`] are routed
+/// based on their [`ProjectKeyPair`].
+#[derive(Debug, Clone)]
+pub struct PartitionedEnvelopeBuffer {
+    buffers: Arc<Vec<ObservableEnvelopeBuffer>>,
+    partitions: u32,
+}
```

Comment:
If we don't need to store `partitions` (see question below), I would simplify this to `struct PartitionedEnvelopeBuffer(Vec<ObservableEnvelopeBuffer>)`. `ObservableEnvelopeBuffer` is already a reference type, and cloning the `Vec` a couple of times should not be that bad.